src/main/java/io/cryostat/util/GsonRuleAdapter.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright The Cryostat Authors
+ *
+ * The Universal Permissive License (UPL), Version 1.0
+ *
+ * Subject to the condition set forth below, permission is hereby granted to any
+ * person obtaining a copy of this software, associated documentation and/or data
+ * (collectively the "Software"), free of charge and under any and all copyright
+ * rights in the Software, and any and all patent rights owned or freely
+ * licensable by each licensor hereunder covering either (i) the unmodified
+ * Software as contributed to or provided by such licensor, or (ii) the Larger
+ * Works (as defined below), to deal in both
+ *
+ * (a) the Software, and
+ * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+ * one is included with the Software (each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.util;
+
+import java.lang.reflect.Type;
+
+import io.cryostat.rules.Rule;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+public class GsonRuleAdapter implements JsonDeserializer<Rule> {
+
+    @Override
+    public Rule deserialize(JsonElement json, Type typeOf, JsonDeserializationContext context)
+            throws IllegalArgumentException {
+
+        JsonObject jsonObject = json.getAsJsonObject();
+        Rule rule = new Gson().fromJson(jsonObject, Rule.class);
```

Comment:
I'd like to sanitize the `name` field, except `sanitizeRuleName()` is private and also has a `fixme`. What will the sanitizer ideally look like? Could it be public somehow to do something like this?

```
public Rule deserialize(JsonElement json, Type typeOf, JsonDeserializationContext context)
            throws IllegalArgumentException {

        JsonObject jsonObject = json.getAsJsonObject();

        String name = Rule.Attribute.NAME.getSerialKey();
        jsonObject.add(name, Rule.sanitizeRuleName(jsonObject.get(name)));

        Rule rule = new Gson().fromJson(jsonObject, Rule.class);
        rule.validate();
        return rule;
    }
```

Reply:
Sure, something like that would definitely work. There's no hard requirement for the method to be (package-)private if there is another good place that needs access to it, and this deserializer is a good place.

The FIXME is about doing more than just replacing whitespace characters with underscores, which is all the sanitization currently does. It should probably also disallow `.jfr` from appearing within a rule name for example.

Alternatively, the rule name validation could be done by the POST handler that processes the user's request to create a Rule definition. If the form or JSON document provided contains an invalid rule name then a `400` response is sent, ideally with some information about why the rule name failed validation. Then it's up to the user/client to try again with a valid name, rather than for the Rule implementation to try to sanitize the name it's given and turn it into something acceptable.

Reply:
Also, btw, it works fine for testing right now, but I don't think a Gson adapter should invoke `new Gson().fromJson()` to perform the actual deserialization. Ideally the adapter should perform this deserialization on its own, or at worst there could be a static Gson instance that is shared and reused for this purpose.